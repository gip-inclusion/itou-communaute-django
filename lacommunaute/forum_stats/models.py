from django.db import models

from lacommunaute.forum_stats.enums import Period


class StatQuerySet(models.QuerySet):
    def current_month_datas(self):
        qs = Stat.objects.filter(period="month").order_by("-date")

        if qs.exists():
            return Stat.objects.filter(date=qs.first().date, period=Period.MONTH).values("name", "value", "date")

        return Stat.objects.none()


class Stat(models.Model):
    name = models.CharField(max_length=30, verbose_name="Nom")
    date = models.DateField(verbose_name="Date")
    value = models.IntegerField(verbose_name="Valeur")
    period = models.CharField(max_length=10, verbose_name="PÃ©riode", choices=Period.choices)

    objects = models.Manager()

    class Meta:
        verbose_name = "Stat"
        verbose_name_plural = "Stats"
        ordering = ["date", "period", "name"]
        unique_together = ("name", "date", "period")

    def __str__(self):
        return f"{self.name} - {self.date} - {self.period}"

    objects = StatQuerySet().as_manager()
