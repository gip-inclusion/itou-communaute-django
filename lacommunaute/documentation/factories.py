import factory
from faker import Faker

from lacommunaute.documentation.models import Category, Document, DocumentRating


faker = Faker()


class AbstractDocumentationFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("name")
    description = factory.Faker("sentence", nb_words=100)
    short_description = factory.Faker("sentence", nb_words=10)
    image = factory.django.ImageField(filename="banner.jpg")


class CategoryFactory(AbstractDocumentationFactory):
    class Meta:
        model = Category

    class Params:
        for_snapshot = factory.Trait(
            name="Test Category", description="Test description", short_description="Test short description"
        )


class DocumentFactory(AbstractDocumentationFactory):
    category = factory.SubFactory(CategoryFactory)

    class Meta:
        model = Document

    class Params:
        for_snapshot = factory.Trait(
            name="Test Document",
            description="Test description",
            short_description="Test short description",
            category=factory.SubFactory(CategoryFactory, for_snapshot=True),
        )

    @factory.post_generation
    def with_tags(self, create, extracted, **kwargs):
        if not create or not extracted:
            return

        if isinstance(extracted, list):
            for tag in extracted:
                self.tags.add(tag)


class DocumentRatingFactory(factory.django.DjangoModelFactory):
    document = factory.SubFactory(DocumentFactory)
    rating = factory.Faker("random_int", min=1, max=5)

    class Meta:
        model = DocumentRating
        skip_postgeneration_save = True

    @factory.post_generation
    def set_created(self, create, extracted, **kwargs):
        if extracted:
            self.created = extracted
            self.save()
