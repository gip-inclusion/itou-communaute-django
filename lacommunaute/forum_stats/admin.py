from django.contrib import admin

from lacommunaute.forum_stats.models import SearchCollectionPeriod, SearchQuery, Stat


class StatAdmin(admin.ModelAdmin):
    list_display = ("name", "date", "value", "period")
    list_filter = ("name", "date", "period")


class SearchCollectionPeriodAdmin(admin.ModelAdmin):
    list_display = ("name", "start_date", "end_date")


class SearchQueryAdmin(admin.ModelAdmin):
    list_display = ("period", "nb_visits")


admin.site.register(Stat, StatAdmin)
admin.site.register(SearchCollectionPeriod, SearchCollectionPeriodAdmin)
admin.site.register(SearchQuery, SearchQueryAdmin)
