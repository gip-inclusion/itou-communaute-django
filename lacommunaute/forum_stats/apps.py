from django.apps import AppConfig


class ForumStatsAppConfig(AppConfig):
    label = "forum_stats"
    name = "lacommunaute.forum_stats"
    verbose_name = "Forum Stats"
    verbose_name_plural = "Forum Stats"
