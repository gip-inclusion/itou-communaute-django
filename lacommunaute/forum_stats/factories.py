import factory
import factory.django

from lacommunaute.forum_stats.enums import Period
from lacommunaute.forum_stats.models import Stat


class StatFactory(factory.django.DjangoModelFactory):
    date = factory.Faker("date")
    name = factory.Faker("word")
    period = Period.DAY
    value = factory.Faker("pyint")

    class Meta:
        model = Stat
