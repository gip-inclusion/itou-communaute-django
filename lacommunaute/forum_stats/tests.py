from dateutil.relativedelta import relativedelta
from django.db import IntegrityError
from django.test import TestCase
from django.utils import timezone
from django.utils.timezone import localdate

from lacommunaute.forum_stats.enums import Period
from lacommunaute.forum_stats.factories import StatFactory
from lacommunaute.forum_stats.models import Stat


class StatModelTest(TestCase):
    def test_uniqueness(self):
        date = timezone.now()
        name = "nb_unique_contributors"
        period = Period.DAY
        StatFactory(name=name, date=date, period=period)
        with self.assertRaises(IntegrityError):
            StatFactory(name=name, date=date, period=period)


class StatQuerySetTest(TestCase):
    def test_ordering(self):
        today = localdate()
        stat = StatFactory(period=Period.MONTH, date=today)
        StatFactory(period=Period.DAY, date=today)
        StatFactory(period=Period.MONTH, date=today - relativedelta(months=1))

        qs = Stat.objects.current_month_datas()
        self.assertEqual(list(qs), [{"name": stat.name, "value": stat.value, "date": today}])

    def test_empty_dataset(self):
        self.assertEqual(Stat.objects.current_month_datas().count(), 0)
