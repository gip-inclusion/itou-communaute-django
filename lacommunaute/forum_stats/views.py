import logging

from dateutil.relativedelta import relativedelta
from django.db.models import CharField
from django.db.models.functions import Cast
from django.utils import timezone
from django.utils.dateformat import format
from django.utils.timezone import localdate
from django.views.generic.base import TemplateView

from lacommunaute.forum_stats.models import Stat
from lacommunaute.utils.json import extract_values_in_list
from lacommunaute.utils.math import percent


logger = logging.getLogger(__name__)


class StatistiquesPageView(TemplateView):
    template_name = "forum_stats/statistiques.html"

    def get_funnel_data(self):
        qs = Stat.objects.current_month_datas()

        stats = {
            "period": None,
            "nb_uniq_visitors": 0,
            "nb_uniq_active_visitors": 0,
            "nb_uniq_engaged_visitors": 0,
        }

        if qs.filter(name="nb_uniq_visitors").exists():
            stats["period"] = format(qs.get(name="nb_uniq_visitors")["date"], "F Y")
            stats["nb_uniq_visitors"] = qs.get(name="nb_uniq_visitors")["value"]

        if qs.filter(name="nb_uniq_active_visitors").exists():
            stats["nb_uniq_active_visitors"] = qs.get(name="nb_uniq_active_visitors")["value"]

        if qs.filter(name="nb_uniq_engaged_visitors").exists():
            stats["nb_uniq_engaged_visitors"] = qs.get(name="nb_uniq_engaged_visitors")["value"]

        stats["activation_percent"] = percent(stats["nb_uniq_active_visitors"], stats["nb_uniq_visitors"])
        stats["engagement_percent"] = percent(stats["nb_uniq_engaged_visitors"], stats["nb_uniq_active_visitors"])
        return stats

    def get_daily_stats(self):
        indicator_names = [
            "nb_uniq_visitors",
            "nb_uniq_active_visitors",
            "nb_uniq_engaged_visitors",
        ]
        after_date = timezone.now() - timezone.timedelta(days=90)
        datas = (
            Stat.objects.filter(period="day", name__in=indicator_names, date__gte=after_date)
            .values("name", "value")
            .annotate(date=Cast("date", CharField()))
        )
        return extract_values_in_list(datas, indicator_names)

    def get_monthly_visitors(self):
        indicator_names = ["nb_uniq_visitors_returning"]
        datas = (
            Stat.objects.filter(period="month", name__in=indicator_names)
            .values("name", "value")
            .annotate(date=Cast("date", CharField()))
        )
        return extract_values_in_list(datas, indicator_names)

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        context["stats"] = self.get_daily_stats()
        context["impact"] = self.get_monthly_visitors()
        context = {**context, **self.get_funnel_data()}

        return context


class MonthlyVisitorsViews(TemplateView):
    template_name = "forum_stats/monthly_visitors.html"

    def get_monthly_visitors(self):
        indicator_names = [
            "nb_uniq_visitors",
            "nb_uniq_active_visitors",
            "nb_uniq_engaged_visitors",
            "nb_uniq_visitors_returning",
        ]
        datas = (
            Stat.objects.filter(
                period="month", name__in=indicator_names, date__gt=localdate() - relativedelta(months=9)
            )
            .values("name", "value")
            .annotate(date=Cast("date", CharField()))
        )
        return extract_values_in_list(datas, indicator_names)

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        context["monthly_visitors"] = self.get_monthly_visitors()
        return context
