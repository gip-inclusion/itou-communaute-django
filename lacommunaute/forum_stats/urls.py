from django.urls import path

from lacommunaute.forum_stats.views import MonthlyVisitorsViews, StatistiquesPageView


app_name = "forum_stats"

urlpatterns = [
    path("", StatistiquesPageView.as_view(), name="statistiques"),
    path("monthly_visitors/", MonthlyVisitorsViews.as_view(), name="monthly_visitors"),
]
