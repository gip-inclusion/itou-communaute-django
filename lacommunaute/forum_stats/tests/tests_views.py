from dateutil.relativedelta import relativedelta
from django.test import TestCase
from django.urls import reverse
from django.utils import timezone
from django.utils.dateformat import format
from django.utils.timezone import localdate
from faker import Faker
from machina.core.loading import get_class
from pytest_django.asserts import assertContains

from lacommunaute.forum_stats.enums import Period
from lacommunaute.forum_stats.factories import StatFactory
from lacommunaute.utils.math import percent


faker = Faker()
assign_perm = get_class("forum_permission.shortcuts", "assign_perm")


class StatistiquesPageTest(TestCase):
    def test_context_data(self):
        url = reverse("forum_stats:statistiques")
        date = timezone.now()
        names = ["nb_uniq_engaged_visitors", "nb_uniq_visitors", "nb_uniq_active_visitors"]
        for name in names:
            StatFactory(name=name, date=date)
        undesired_period_stat = StatFactory(
            period=Period.WEEK, date=date - timezone.timedelta(days=7), name="nb_uniq_engaged_visitors"
        )
        undesired_date_stat = StatFactory(
            period=Period.DAY, date=date - timezone.timedelta(days=91), name="nb_uniq_engaged_visitors"
        )

        response = self.client.get(url)
        self.assertEqual(response.status_code, 200)
        self.assertTemplateUsed(response, "forum_stats/statistiques.html")

        # expected values
        self.assertIn("stats", response.context)
        self.assertIn("date", response.context["stats"])
        self.assertIn("nb_uniq_engaged_visitors", response.context["stats"])
        self.assertIn("nb_uniq_visitors", response.context["stats"])
        self.assertIn("nb_uniq_active_visitors", response.context["stats"])
        self.assertEqual(response.context["stats"]["date"][0], date.strftime("%Y-%m-%d"))

        # undesired values
        self.assertNotIn(undesired_period_stat.date.strftime("%Y-%m-%d"), response.context["stats"]["date"])
        self.assertNotIn(undesired_date_stat.date.strftime("%Y-%m-%d"), response.context["stats"]["date"])

    def test_month_datas_in_context(self):
        today = localdate()
        url = reverse("forum_stats:statistiques")

        # no data
        response = self.client.get(url)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.context["period"], None)
        self.assertEqual(response.context["nb_uniq_visitors"], 0)
        self.assertEqual(response.context["nb_uniq_active_visitors"], 0)
        self.assertEqual(response.context["nb_uniq_engaged_visitors"], 0)
        self.assertEqual(response.context["activation_percent"], 0)
        self.assertEqual(response.context["engagement_percent"], 0)

        # undesired data
        StatFactory(name="nb_uniq_engaged_visitors", period=Period.DAY, date=today)
        StatFactory(name=faker.word(), period=Period.MONTH, date=today)
        response = self.client.get(url)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.context["period"], None)
        self.assertEqual(response.context["nb_uniq_visitors"], 0)
        self.assertEqual(response.context["nb_uniq_active_visitors"], 0)
        self.assertEqual(response.context["nb_uniq_engaged_visitors"], 0)
        self.assertEqual(response.context["activation_percent"], 0)
        self.assertEqual(response.context["engagement_percent"], 0)

        uniq_visitors = StatFactory(name="nb_uniq_visitors", period=Period.MONTH, date=today)
        uniq_active_visitors = StatFactory(name="nb_uniq_active_visitors", period=Period.MONTH, date=today)
        uniq_engaged_visitors = StatFactory(name="nb_uniq_engaged_visitors", period=Period.MONTH, date=today)
        response = self.client.get(url)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.context["period"], format(today, "F Y"))
        self.assertEqual(response.context["nb_uniq_visitors"], uniq_visitors.value)
        self.assertEqual(response.context["nb_uniq_active_visitors"], uniq_active_visitors.value)
        self.assertEqual(response.context["nb_uniq_engaged_visitors"], uniq_engaged_visitors.value)
        self.assertEqual(
            response.context["activation_percent"], percent(uniq_active_visitors.value, uniq_visitors.value)
        )
        self.assertEqual(
            response.context["engagement_percent"], percent(uniq_engaged_visitors.value, uniq_active_visitors.value)
        )

    def test_impact_in_context_data(self):
        url = reverse("forum_stats:statistiques")
        today = localdate()
        empty_res = {"date": [], "nb_uniq_visitors_returning": []}

        # no data
        response = self.client.get(url)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.context["impact"], empty_res)

        # undesired data
        StatFactory(name="nb_uniq_visitors_returning", period=Period.DAY, date=today)
        StatFactory(name=faker.word(), period=Period.MONTH, date=today)
        response = self.client.get(url)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.context["impact"], empty_res)

        # desired data
        StatFactory(name="nb_uniq_visitors_returning", period=Period.MONTH, date=today, value=1)
        response = self.client.get(url)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.context["impact"]["date"][0], today.strftime("%Y-%m-%d"))
        self.assertEqual(response.context["impact"]["nb_uniq_visitors_returning"][0], 1)

    def test_navigation(self):
        url = reverse("forum_stats:statistiques")
        response = self.client.get(url)
        self.assertContains(response, "<a href=/statistiques/monthly_visitors/>")


class TestMonthlyVisitorsView:
    def test_context_data(self, client, db):
        url = reverse("forum_stats:monthly_visitors")
        today = localdate()
        empty_res = {
            "date": [],
            "nb_uniq_visitors": [],
            "nb_uniq_active_visitors": [],
            "nb_uniq_engaged_visitors": [],
            "nb_uniq_visitors_returning": [],
        }

        # no data
        response = client.get(url)
        assert response.status_code == 200
        assertContains(response, "Utilisateurs uniques mensuels")
        assert response.context["monthly_visitors"] == empty_res

        # undesired data
        StatFactory(name="nb_uniq_visitors_returning", period=Period.DAY, date=today)
        StatFactory(name=faker.word(), period=Period.MONTH, date=today)
        StatFactory(name="nb_uniq_visitors", period=Period.MONTH, date=today - relativedelta(months=9), value=1)
        response = client.get(url)
        assert response.status_code == 200
        assert response.context["monthly_visitors"] == empty_res

        # expected data
        StatFactory(name="nb_uniq_visitors_returning", period=Period.MONTH, date=today, value=2)
        StatFactory(name="nb_uniq_visitors", period=Period.MONTH, date=today - relativedelta(months=8), value=10)
        response = client.get(url)
        assert response.status_code == 200
        assert response.context["monthly_visitors"] == {
            "date": [(today - relativedelta(months=8)).strftime("%Y-%m-%d"), today.strftime("%Y-%m-%d")],
            "nb_uniq_visitors": [10],
            "nb_uniq_active_visitors": [],
            "nb_uniq_engaged_visitors": [],
            "nb_uniq_visitors_returning": [2],
        }

    def test_navigation(self, client, db):
        url = reverse("forum_stats:monthly_visitors")
        response = client.get(url)
        assert response.status_code == 200
        assertContains(response, '<a href="/statistiques/">retour vers la page statistiques</a>')
