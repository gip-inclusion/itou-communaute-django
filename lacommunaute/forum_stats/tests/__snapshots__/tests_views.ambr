# serializer version: 1
# name: TestStatistiquesPageView.test_dsp_count[dsp]
  '''
  <div class="s-section__row row" id="daily_dsp">
                  <div class="s-section__col col-12">
                      <div class="c-box mb-3 mb-md-5">
                          <h2>10 Diagnostiques Parcours IAE</h2>
                          <p>réalisés sur la communauté de l'inclusion depuis le 24 février 2024</p>
                          <small>
                              <a href="/statistiques/dsp/">Voir les détails</a>
                          </small>
                      </div>
                  </div>
              </div>
  '''
# ---
